cmd/collectors/zapiperf/plugins/vscan/vscan.go

```diff
@@ -0,0 +1,151 @@
+package vscan
+
+import (
+	"goharvest2/cmd/poller/plugin"
+	"goharvest2/pkg/matrix"
+	"strconv"
+	"strings"
+)
+
+type Vscan struct {
+	*plugin.AbstractPlugin
+}
+
+func New(p *plugin.AbstractPlugin) plugin.Plugin {
+	return &Vscan{AbstractPlugin: p}
+}
+
+func (v *Vscan) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
+	// defaults plugin options
+	isPerScanner := true
+
+	if s := v.Params.GetChildContentS("metricsPerScanner"); s != "" {
+		if parseBool, err := strconv.ParseBool(s); err == nil {
+			isPerScanner = parseBool
+		}
+	}
+	v.Logger.Debug().Bool("isPerScanner", isPerScanner).Msg("Vscan options")
+
+	v.addSvmAndScannerLabels(data)
+	if !isPerScanner {
+		return nil, nil
+	}
+
+	return v.aggregatePerScanner(data)
+}
+
+func (v *Vscan) addSvmAndScannerLabels(data *matrix.Matrix) {
+	for _, instance := range data.GetInstances() {
+		ontapName := instance.GetLabel("instance_uuid")
+		// colon separated list of fields
+		// vs_test4   :    2.2.2.2   :    umeng-aff300-05
+		//  svm       :     scanner  :    node
+		if split := strings.Split(ontapName, ":"); len(split) >= 3 {
+			instance.SetLabel("svm", split[0])
+			instance.SetLabel("scanner", split[1])
+			instance.SetLabel("node", split[2])
+		}
+	}
+}
+
+func (v *Vscan) aggregatePerScanner(data *matrix.Matrix) ([]*matrix.Matrix, error) {
+	// When isPerScanner=true, Harvest 1.6 uses this form:
+	// netapp.perf.dev.nltl-fas2520.vscan.scanner.10_64_30_62.scanner_stats_pct_mem_used 18 1501765640
+
+	// These counters are per scanner and need averaging:
+	// 		scanner_stats_pct_cpu_used
+	// 		scanner_stats_pct_mem_used
+	// 		scanner_stats_pct_network_used
+	// These counters need to be summed:
+	// 		scan_request_dispatched_rate
+
+	// create per scanner instance cache
+	cache := data.Clone(false, true, false)
+	cache.UUID += ".Vscan"
+
+	for _, i := range data.GetInstances() {
+		scanner := i.GetLabel("scanner")
+		if cache.GetInstance(scanner) == nil {
+			s, _ := cache.NewInstance(scanner)
+			s.SetLabel("scanner", scanner)
+		}
+		i.SetExportable(false)
```

Comment:
Why do we need to SetExportable(false)

Reply:
if not set, the clone matrix and the original are exported.
set, only the clone is exported